src/particle/corsika/corsika7id.py

```diff
@@ -0,0 +1,168 @@
+"""
+
+
+from __future__ import annotations
+
+import csv
+from typing import TypeVar
+
+from .. import data
+from ..exceptions import MatchingIDNotFound
+from ..pdgid import PDGID
+
+Self = TypeVar("Self", bound="Corsika7ID")
+
+
+with data.basepath.joinpath("pdgid_to_corsika7id.csv").open() as _f:
+    _bimap = {
+        int(v["CORSIKA7ID"]): int(v["PDGID"])
+        for v in csv.DictReader(line for line in _f if not line.startswith("#"))
+    }
+
+# Some Corsika ID's are not really particles
+_non_particles = {
+    71: "η → γγ",
+    72: "η → 3π◦",
+    73: "η → π+π−π◦",
+    74: "η → π+π−γ",
+    75: "μ+ add. info.",
+    76: "μ− add. info.",
+    85: "decaying μ+ at start",
+    86: "decaying μ− at start",
+    95: "decaying μ+ at end",
+    96: "decaying μ− at end",
+}
+
+
+class Corsika7ID(int):
+    """
+    Holds a Corsika7 ID.
+
+    Examples
+    --------
+    >>> cid = Corsika7ID(6)
+
+    >>> from particle import Particle
+    >>> p = Particle.from_pdgid(cid.to_pdgid())
+
+    >>> (p,) = Particle.finditer(pdgid=cid.to_pdgid())
+    >>> p.name
+    'mu-'
+    """
+
+    @classmethod
+    def from_pdgid(cls: type[Self], pdgid: int) -> Self:
+        """
+        Constructor from a PDGID.
+        """
+        for k, v in _bimap.items():
+            if v == pdgid:
+                return cls(k)
+        raise MatchingIDNotFound(f"Non-existent Corsika7ID for input PDGID {pdgid}!")
+
+    @classmethod
+    def from_particle_description(
+        cls: type[Self], particle_description: int
+    ) -> tuple[Self, bool]:
+        """
+        Constructor from the particle description returned by Corsika7
+        in the particle data sub-block, mother particle data sub-block or
+        the grandmother particle data sub-block.
+
+        Returns
+        -------
+        Corsika7ID: The Corsika7 id
+        bool:       If the particle is a (grand)motherparticle.
+        """
+        cid = abs(particle_description) // 1000
+        ismother = particle_description < 0
+
+        if cls._is_non_particle_id(cid):
+            return cls(cid), ismother
+
+        # this catches the case, of nuclei with no known PDGid
+        if cid >= 200 and cid < 5699:
+            return cls(cid), ismother
+
+        if cid in _bimap:
+            return cls(cid), ismother
+
+        raise MatchingIDNotFound(
+            f"Non-existent Corsika7ID for particle_description {particle_description}!"
+        )
+
+    @classmethod
+    def _is_non_particle_id(cls: type[Self], id: int) -> bool:
+        """
+        returns True if the id is a valid id, but not a particle, False otherwise.
+        """
+        return id in _non_particles or id // 1000 == 8888 or id == 9900
+
+    def is_particle(self) -> bool:
+        """
+        Returns if the corsikaid really belongs to a particle, since some are for example additional information.
+        """
+        iid = int(self)
+
+        return not self._is_non_particle_id(iid)
+
+    def name(self) -> str:
```

Comment:
Also a function that deserves several examples for what is possible, especially given the comments and possible raise.